Fix a failing test in NetworkPolicy e2e

tnqn opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
The test "should allow ingress access from updated pod" fails regardless
of which CNI plugin is enabled. It's because the test assumes the client
Pod can recheck connectivity after updating its label, but the client
won't restart after the first failure, so the second check will always
fail. The PR creates a client Pod with OnFailure RestartPolicy to fix it.

In addition to the above test that checks rule selector takes effect on
updated client pod, the PR adds a test "should deny ingress access to
updated pod" to ensure network policy selector can take effect on updated
server pod.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #85908

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

tnqn committed:
Fix updated pod NetworkPolicy e2e test

The test "should allow ingress access from updated pod" fails regardless
of which CNI plugin is enabled. It's because the test assumes the client
Pod can recheck connectivity after updating its label, but the client
won't restart after the first failure, so the second check will always
fail. The PR creates a client Pod with OnFailure RestartPolicy to fix it.

In addition to the above test that checks rule selector takes effect on
updated client pod, the PR adds a test "should deny ingress access to
updated pod" to ensure network policy selector can take effect on updated
server pod.

tnqn commented:
/assign @danwinship

matthyx commented:
/ok-to-test
/cc
/assign @liggitt

liggitt commented:
sig-network owns review of this test

matthyx commented:
> sig-network owns review of this test

sorry

tnqn commented:
/retest

tnqn commented:
/retest

tnqn commented:
/retest

jayunit100 commented:
cc @andrewsykim  .... i like that this fixes network policy tests, can you confirm that we're not altering the semantics of the test in a way that can cause a regression ?   Im looking into this as well no.

andrewsykim commented on test/e2e/network/network_policy.go:
```diff
@@ -821,49 +821,39 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
 			policy := &networkingv1.NetworkPolicy{
 				ObjectMeta: metav1.ObjectMeta{
-					Name: "allow-pod-b-via-pod-selector",
+					Name: "deny-all",
```
A better name might be `deny-ingress-via-isolated-label-selector`

tnqn replied:
Done

tnqn commented:
Thanks @andrewsykim @jayunit100 for reviewing and the suggestion.

andrewsykim commented on test/e2e/network/network_policy.go:
```diff
@@ -821,49 +821,39 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
+			// Client cannot connect to service after updating the server pod's labels to match the network policy.
+			ginkgo.By(fmt.Sprintf("Updating server pod %s to be selected by network policy %s.", podServer.Name, policy.Name))
+			updateNetworkClientPodLabel(f, f.Namespace, podServer.Name, "add", "/metadata/labels", map[string]string{"isolated": ""})
+			testCannotConnect(f, f.Namespace, "client-a", service, allowedPort)
```
Pod client has `RestartPolicyNever`, so would this also fail since pod will be in `Succeeded` state from previous test case?

tnqn replied:
`testCanConnect` and `testCannotConnect` create new client Pod, check connection, and clean up it. 
https://github.com/kubernetes/kubernetes/blob/52d7614a8ca5b8aebc45333b6dc8fbf86a5e7ddf/test/e2e/network/network_policy.go#L1386-L1396
Using same pod name `client-a` is to emphasize the behavior changes by server pod's label change only. The two steps actually use two client Pods having same name and labels.

andrewsykim replied:
So this is no longer "ingress access to updated pod", but "ingress access to new pod" then? I think we still want the updated pod test case.

tnqn replied:
It's still updated pod, but just updated server Pod.
As the PR description explains, all network policy tests are using the utils "testCannotConnect" and "testCanConnect" which creates a client Pod designed for **one-shot** check without restarting, it's not convenient to test twice with a single client Pod (that's also the mistake the original test made), unless we write another util just for this case.
I think essentially the updated pod test case wants to ensure network policy can be enabled on existing Pods after updating their labels to match them, so it should make sense to verify it on server Pod, as even the original test just verified it on client Pod. Does it make sense?

andrewsykim replied:
I see, so if we are creating new client pod every time and only the server pod gets updated labels, it seems like we can still test the `deny` -> `allow` test case right and we don't really need to flip the order? I think we could benefit from having both:
* deny -> allow by updating server pod labels with new client pod for each connection
* allow -> deny by updating server pod labels with new client pod for each connection

tnqn replied:
I updated the order because the server Pod and its service is created in `BeforeEach`, having only a fixed label `pod-name: server` which is also used by its service to select backends. So it's not convenient to first create a deny policy which selects the server Pod, and then update the server Pod to not match it as updating label "pod-name: server" will make service not work.
To have both tests, I can append one more step to remove the "isolated" label and call `testCanConnect` again, it would be the reversed version of your proposal, what do you think?

lzang replied:
This looks like a different test case to me. The original one is to change the network policy ingress rule selector, so a client that couldn't connect can connect now. Your test case is to change network policy selector, so a server that was protected by the policy become not protected. Maybe we can fix the original one by creating a new client pod and create a new test case for this change? Creating a new client pod is not ideal to test policy change but the enforcement is on the server side for both cases anyway.

tnqn replied:
Thanks @lzang for reviewing. I restored the original test and fixed it by creating the client pod with `OnFailure` RestartPolicy and updating its label after failure is detected as creating a new client pod may make the test case same as some existing tests which check pods with different labels will have different behaviors, like 
https://github.com/kubernetes/kubernetes/blob/29b09c7fb1a8d38ab9df4873553cc0b8d97aae95/test/e2e/network/network_policy.go#L147-L152
I created a new test case to ensure network policy selector can take effect on updated server pod as you suggested.

lzang replied:
The existing one doesn't have policy update, so I don't think they are the same, but I am fine if the new approach by modifying the restart policy could work.

bowei commented:
cc: @lzang

lzang commented on test/e2e/network/network_policy.go:
```diff
@@ -821,49 +821,39 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
+
+			// Client cannot connect to service after updating the server pod's labels to match the network policy.
+			ginkgo.By(fmt.Sprintf("Updating server pod %s to be selected by network policy %s.", podServer.Name, policy.Name))
+			updateNetworkClientPodLabel(f, f.Namespace, podServer.Name, "add", "/metadata/labels", map[string]string{"isolated": ""})
```
Use updateNetworkClientPodLabel to change a server pod label is confusing. Could you rename the function to a generic one?

tnqn replied:
I didn't notice the function name, thanks for catching it, done.

bowei commented:
/assign @lzang

matthyx commented:
/assign @lzang

tnqn commented:
/retest

tnqn commented:
/retest

lzang commented on test/e2e/network/network_policy.go:
```diff
@@ -852,20 +852,75 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
 			ginkgo.By(fmt.Sprintf("Updating client pod %s that should successfully connect to %s.", podClient.Name, service.Name))
-			podClient = updateNetworkClientPodLabel(f, f.Namespace, podClient.Name, "replace", "/metadata/labels", map[string]string{})
+			podClient = updatePodLabel(f, f.Namespace, podClient.Name, "replace", "/metadata/labels", map[string]string{})
 			checkConnectivity(f, f.Namespace, podClient, service)
```
If checkConnectivity fails, who will clean up the pod as it would keep restarting on failure?

tnqn replied:
`checkConnectivity` has timeout (3 min):
https://github.com/kubernetes/kubernetes/blob/ae95a4bfcac12bf6b19e2d5acf6404359b1a8c3b/test/e2e/framework/pod/wait.go#L356-L357
The test will fail if timeout occurs and the Pod will be cleaned up by the deferred delete function.

lzang commented on test/e2e/network/network_policy.go:
```diff
@@ -1415,28 +1470,7 @@ func checkConnectivity(f *framework.Framework, ns *v1.Namespace, podClient *v1.P
-		for _, p := range podsInNS {
-			pods = append(pods, fmt.Sprintf("Pod: %s, Status: %s\n", p.Name, p.Status.String()))
-		}
+		pods, policies, logs := collectPodsAndNetworkPolicies(f, podClient)
```
Thanks for cleaning this up!

tnqn commented on test/e2e/network/network_policy.go:
```diff
@@ -852,20 +852,75 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
 			}()
-			checkNoConnectivity(f, f.Namespace, podClient, service)
+			// Check Container exit code as restartable Pod's Phase will be Running even when container fails.
+			err = e2epod.WaitForPodCondition(f.ClientSet, f.Namespace.Name, podClient.Name, "terminated", framework.PodStartTimeout, func(pod *v1.Pod) (bool, error) {
```
Done

tnqn commented:
/retest

tnqn commented:
@andrewsykim @bowei @lzang I think I have addressed all comments, fixed the original updated pod (matching rule selector) case, and added a new one for updated server pod (matching network policy selector), could you take another look at this?

lzang commented:
/lgtm

matthyx commented:
/lgtm

tnqn commented:
Resolved conflicts.
@andrewsykim @bowei @lzang @matthyx could you review again and approve it if the fix makes sense to you?

bowei commented:
/approve

bowei commented:
/lgtm
